tests/components/modbus/conftest.py

```diff
@@ -37,63 +32,101 @@ def __init__(self, register_words):
+        return_value=mock_sync,
+    ), mock.patch(
+        "homeassistant.components.modbus.modbus.ModbusUdpClient", return_value=mock_sync
+    ):
+
+        # Setup inputs for the sensor
+        read_result = ReadResult(register_words)
+        mock_sync.read_coils.return_value = read_result
+        mock_sync.read_discrete_inputs.return_value = read_result
+        mock_sync.read_input_registers.return_value = read_result
+        mock_sync.read_holding_registers.return_value = read_result
+
+        # mock timer and add modbus platform with devices (new config)
+        now = dt_util.utcnow()
+        with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+            if method_discovery:
+                # setup modbus which in turn does setup for the devices
+                if config_device is not None:
+                    config_modbus[DOMAIN].update(
+                        {array_name_discovery: [{**config_device}]}
+                    )
+                await async_load_platform(
+                    hass, DOMAIN, DOMAIN, config_modbus[DOMAIN], config_modbus
+                )
+            else:
+                # first add modbus platform using old config
+                assert await async_setup_component(hass, DOMAIN, config_modbus)
+
+                # setup component old style
+                if config_device is not None:
+                    config_device = {
+                        entity_domain: {
+                            CONF_PLATFORM: DOMAIN,
+                            array_name_old_config: [
+                                {
+                                    **config_device,
+                                }
+                            ],
+                        }
+                    }
+                    if scan_interval is not None:
+                        config_device[entity_domain][CONF_SCAN_INTERVAL] = scan_interval
+                    await hass.async_block_till_done()
+                    assert await async_setup_component(
+                        hass, entity_domain, config_device
+                    )
+            await hass.async_block_till_done()
+
+        assert DOMAIN in hass.data
+        if config_device is not None:
+            entity_id = f"{entity_domain}.{device_name}"
+            device = hass.states.get(entity_id)
+            if device is None:
+                pytest.fail("CONFIG failed, see output")
+        if check_config_only:
+            return
+
+        # Trigger update call with time_changed event
+        now = now + timedelta(seconds=scan_interval + 60)
+        with mock.patch("homeassistant.helpers.event.dt_util.utcnow", return_value=now):
+            async_fire_time_changed(hass, now)
+            await hass.async_block_till_done()
+
+        # Check state
+        entity_id = f"{entity_domain}.{device_name}"
+        state = hass.states.get(entity_id).state
+        assert state == expected
```

Comment:
It's more readable if we keep the assertion in the tests and just have the setup code in this helper function.

Reply:
Agreed and done. On top of that I have added a cover function base_config_test, so it is very clear in the test functions which functionality is wanted (test or just configuration check).

Hope that solves your comments.